tools/compile_et/compile_et.1
.\" Copyright (c) 1988 Massachusetts Institute of Technology,
.\" Student Information Processing Board.  All rights reserved.
.\"
.\" $OpenBSD: compile_et.1,v 1.6 1999/07/07 10:50:07 aaron Exp $
.\" $Header: /cvsroot/posix2/posix2/tools/compile_et/compile_et.1,v 1.2 2000/08/14 16:26:07 amai Exp $
.\"
.Dd	November 22, 1988
.Dt COMPILE_ET 1
.Os
.Sh NAME
.Nm compile_et
.Nd error table compiler
.Sh SYNOPSIS
.Nm compile_et
.Ar file
.Sh DESCRIPTION
.Nm
converts a table listing error-code names and associated messages into
a C source file suitable for use with the
.Xr com_err 3
library.
.Pp
The source file name must end with a suffix of ``.et''; the file
consists of a declaration supplying the name (up to four characters
long) of the error-code table:

.Em error_table name

followed by up to 256 entries of the form:

.Em error_code name ,
"
.Em string
"

and a final

.Em end

to indicate the end of the table.
.Pp
The name of the table is used to construct the name of a subroutine
.Em initialize_XXXX_error_table
which must be called in order for the
.Xr com_err 3
library to recognize the error table.
.Pp
The various error codes defined are assigned sequentially increasing
numbers (starting with a large number computed as a hash function of
the name of the table); thus for compatibility it is suggested that
new codes be added only to the end of an existing table, and that no
codes be removed from tables.
.Pp
The names defined in the table are placed into a C header file with
preprocessor directives defining them as integer constants of up to
32 bits in magnitude.
.Pp
A C source file is also generated which should be compiled and linked
with the object files which reference these error codes; it contains
the text of the messages and the initialization subroutine.  Both C
files have names derived from that of the original source file, with
the ``.et'' suffix replaced by ``.c'' and ``.h''.
.Pp
A ``#'' in the source file is treated as a comment character, and all
remaining text to the end of the source line will be ignored.
.Sh BUGS
Since
.Nm
uses a very simple parser based on
.Xr yacc 1 ,
its error recovery leaves much to be desired.
.Sh "SEE ALSO"
.Xr yacc 1 ,
.Xr com_err 3
.Pp
.Rs
.%A Ken Raeburn
.%T "A Common Error Description Library for UNIX"
.Re


ChangeLog
		[$Id: ChangeLog,v 1.34 2000/11/27 18:13:42 arnd Exp $]
 
2000/03/14 LIBREVISION 0.01
	Initial version (arnd)
2000/03/22
	added fixcrcr utility and missing Makefiles in libext/uni (veit)

2000/03/24 LIBREVISION 0.02
	big change: (arnd)
	beautified fixcrcr.c, more docs, added syslog/drand/valloc/dlfcn,
	more headers, added some functions from EMX stat.c, added some
	i387 libm asm funcs, added a math test vector
2000/03/29 
	additions: (arnd)
	signal,resource,time header files and man pages
	Updated CodeRules
	Added Id lines to various files
	Support for itimer and daemon
	Misc fixes

2000/04/03 LIBREVISION 0.03
	additions: (arnd)
	more BSD man2 docs, signal(3) docs
	some missing headers (limits, signal, cdefs, itimer, resource)
	signalx.c was missing in CVS
	popen.c
	fixed a few Makefiles
	more CVSIds in files

2000/04/17 LIBREVISION 0.04 (0.04a)
	additions: (arnd)
	more docs, several sys headers from BSD
	regex support, some more stdlib: a64l, quad, tsearch
	sysconf support
	test for popen
	renamed fixcrcr.exe to fixtxt.exe
2000/04/18 (0.04b)
	additions: (arnd)
	err* stuff, __volatile__ macro fixed
	several minor CVS-Id changes
	msginfo structure renamed (clash with macro)
	removed timext.h
	_getPosixOS2Root cast fixed

2000/04/26 LIBREVISION 0.05
	additions: (arnd)
	perror, vis, man pages
	declare several arrays as pointers rather than []
	various conformance tests for FP and regex
	finally remove timext.h :-)
	set several constants from param.h which where hardcoded before
	init some local variables in regcomp to NULL
	empty getloadavg
	added some assembler math functions (i387math.S) from libgcc
	regexp debug function

2000/05/02 LIBREVISION 0.06
	additions: (arnd, Stefan Neis, Ralf Barkow)
	i387math.S fixes for acos/asin/exp from FreeBSD (neis)
	move fcntl.h after sys/types.h (barkow) and sys/stat.h (arnd)
	man pages: poll,bitstring,cat*,get*,killpg,math,hz,intro9
	i386 specific float.h, ieee.h->ieeefp.h
	nls support,killpg declaration, 
	IPC and SHM fixed,zlib integration
	mmap from BSD removed, nls messages for different languages
	sysconf fixes
	tests for db,shm
2000/05/05
	Math update (Cephes Math Lib), addition of SYS-facitilities.txt
	moved drand
	don't use gcc builtins
	handling of infinity
	test for floatingpoint funcs, consistency test
2000/05/08
	Math update s_tanh.c (neis)
2000/05/12
	Fix to stdio and unistd headers to make C++ happy (neis)

2000/05/15 LIBREVISION 0.07
	additions: ctype, libgen,
	docs for mmap, man5 pages
	tools: the ultimate editor (porting example), pmake
2000/05/17
	Minor update (arnd)
	additions: 4.3 compatible regex, realpath, nice
	more manpages
	HowTo pages for some OS/2 stuff
2000/05/31
	Update rev007b (arnd)
	fixcrcr.c: added check to detect binaries (\0 bytes in text)
	changed librevision format in dll.c
	added several EDM/2 articles and moved html files to docs/html
	some more manpages
	More stuff to FAQ
	renamed some fdlibm functions to __fdlibm
	__STRICT_ANSI__ removed from some defines
	register in prototypes removed
	itimer management improved
	added getsubopt,l64a,mergesort,bm,strlcpy,strmode,strsep
	non-Posix string extensions: strcasecmp etc.
2000/06/13
	math.h _LIB_VERSION var moved (neis)
2000/07/06
	minor changes of defs in math.h and limits.h for LONG LONG support
	(arnd)

Alpha #1 released

2000/07/24
	missing stuff for tools/pmake (arnd)
2000/07/25
	missing newline character added in some headers (amai)
	OpenBSD version of db added (arnd)
	misc fixes: headers, sources, docs (arnd)
2000/07/28
	<poll.h> added, minor updates for docs (amai)

2000/07/29 LIBREVISION 0.09
	lockf() added: not functional yet (see fcntl() in EMX) (arnd)
	timezone stuff added (arnd)
        Minor overall fixed (Makefiles, docs) (amai)

Alpha #2 released

2000/08/06
	Update rev009b
	time.h added, mmap fixes (arnd)
	removed #warnings from some headers - will make autoconf happy (arnd)
2000/08/08
	Update rev009c
	signgam added to libm (arnd)
	plain text version of FAQ removed in favour of HTML version (arnd)
	Makefile fixes (arnd, amai)
	[minor fixes to Makefiles and documentation are not always
	 logged here!]

2000/08/14 LIBREVISION 0.0.10
	misc fixes for libext (move some stuff in according subdirectories,
	dirent stuff, ...)   (arnd)
	Add big tree of utilities to tools/
2000/08/14 LIBREVISION 0.0.11
	few minor fixes for libExt (arnd)
	Modifications for Makefiles, make use of .mk files (arnd)
	"Missing" stuff in tools/ from rev 010 (arnd)
	working copy of rename(); tests for rename() and stat() (amai)
2000/08/15 LIBREVISION 0.0.11a
	few minor fixes for libExt; cleanup of directories (arnd)
	add data files to src/share (arnd)
	misc fixes for tools/ tree; added find(1) (arnd)
2000/08/18 LIBREVISION 0.0.11b
	Misc changes and additions; more files moved location (arnd)
	(In an early state this release was messed up in CVS!)
2000/08/19 LIBREVISION 0.0.12
	Misc stuff as usual. Mainly Makfiles this time,
	introduce two dlls instead of the single one
      change directory structures a bit (arnd)
2000/08/19 LIBREVISION 0.0.12a
2000/08/21 LIBREVISION 0.0.12b
2000/08/23 LIBREVISION 0.0.12c
	Dirent headers, glob* + misc stuff (arnd)
2000/08/26 LIBREVISION 0.0.13
	Build up basic directory structures (arnd)
      Add calendar and rmdir to tools/ as well as some manpages (arnd)
      Fixes/additions to libext
2000/09/09 LIBREVISION 0.0.14
      Add asa and printenv to tools/ as well as some manpages and 
      docs (arnd).
      Fixes/additions to libExt (arnd):
	1) New directory libext/libc/sys for modules that emulate BSD system 
	calls in a safer and more compatible way. Moved several ones.
	2) sys/mm.c enhanced and prepared for later split that will be 
	necessary for linking into a run-time DLL. New function 
	mprotect(). Now mmap() works with MAP_SHARED.
	3) sys/statx.c enhanced: New function statfs(). Several headers 
	adapted: sys/vfs.h, sys/statfs.h, sys/mount.h.
2000/10/11 LIBREVISION 0.0.14a (arnd)
	more changes of mm.c (MAP_SHARED/MAP_FIXED better supported)

tools/compile_et/compile_et.c
/*	$Id: compile_et.c,v 1.2 2000/08/14 16:26:07 amai Exp $	*/

/*
 * Copyright 1986, 1987, 1988
 * by MIT Student Information Processing Board.
 *
 * For copyright info, see "mit-sipb-copyright.h".
 */

#include <stdio.h>
#include <sys/types.h>
#include <sys/file.h>
#include <string.h>
#include <sys/param.h>
#include "compiler.h"

#ifndef __STDC__
#define const
#endif

#ifndef lint
static const char copyright[] =
    "Copyright 1987,1988 by MIT Student Information Processing Board";

static const char rcsid_compile_et_c[] =
    "$Id: compile_et.c,v 1.2 2000/08/14 16:26:07 amai Exp $";
#endif

extern char *gensym();
extern char *current_token;
extern int table_number, current;
char buffer[BUFSIZ];
char *table_name = NULL;
FILE *hfile, *cfile;

/* C library */
extern char *malloc();
extern int errno;

/* lex stuff */
extern FILE *yyin;
extern unsigned lineno;

char * xmalloc (size) unsigned int size; {
    char * p = malloc (size);
    if (!p) {
	perror (whoami);
	exit (1);
    }
    return p;
}

static int check_arg (str_list, arg) char const *const *str_list, *arg; {
    while (*str_list)
	if (!strcmp(arg, *str_list++))
	    return 1;
    return 0;
}

static const char *const debug_args[] = {
    "d",
    "debug",
    0,
};

static const char *const lang_args[] = {
    "lang",
    "language",
    0,
};

static const char *const language_names[] = {
    "C",
    "K&R C",
    "C++",
    0,
};

static const char * const c_src_prolog[] = {
    "static const char * const text[] = {\n",
    0,
};

static const char * const krc_src_prolog[] = {
    "#ifdef __STDC__\n",
    "#define NOARGS void\n",
    "#else\n",
    "#define NOARGS\n",
    "#define const\n",
    "#endif\n\n",
    "static const char * const text[] = {\n",
    0,
};

static const char *const struct_def[] = {
    "struct error_table {\n",
    "    char const * const * msgs;\n",
    "    long base;\n",
    "    int n_msgs;\n",
    "};\n",
    "struct et_list {\n",
    "    struct et_list *next;\n",
    "    const struct error_table * table;\n",
    "};\n",
    "extern struct et_list *_et_list;\n",
    "\n", 0,
};

static const char warning[] =
    "/*\n * %s:\n * This file is automatically generated; please do not edit it.\n */\n";

/* pathnames */
char c_file[MAXPATHLEN];	/* output file */
char h_file[MAXPATHLEN];	/* output */

static void usage () {
    fprintf (stderr, "usage: %s ERROR_TABLE\n",
	     whoami);
    exit (1);
}

static void dup_err (type, one, two) char const *type, *one, *two; {
    fprintf (stderr, "%s: multiple %s specified: `%s' and `%s'\n",
	     whoami, type, one, two);
    usage ();
}

int main(int,char**);
int
main(argc, argv)
	int argc;
	char **argv;
{

	__progname = argv[0];
	_wildcard(&argc, &argv);

    char *p, *ename;
    int len;
    char const * const *cpp;
    int got_language = 0;

    /* argument parsing */
    debug = 0;
    filename = 0;
    whoami = argv[0];
    p = strrchr (whoami, '/');
    if (p)
	whoami = p+1;
    while (argv++, --argc) {
	char *arg = *argv;
	if (arg[0] != '-') {
	    if (filename)
		dup_err ("filenames", filename, arg);
	    filename = arg;
	}
	else {
	    arg++;
	    if (check_arg (debug_args, arg))
		debug++;
	    else if (check_arg (lang_args, arg)) {
		got_language++;
		arg = *++argv, argc--;
		if (!arg)
		    usage ();
		if (language)
		    dup_err ("languanges", language_names[(int)language], arg);
#define check_lang(x,v) else if (!strcasecmp(arg,x)) language = v
		check_lang ("c", lang_C);
		check_lang ("ansi_c", lang_C);
		check_lang ("ansi-c", lang_C);
		check_lang ("krc", lang_KRC);
		check_lang ("kr_c", lang_KRC);
		check_lang ("kr-c", lang_KRC);
		check_lang ("k&r-c", lang_KRC);
		check_lang ("k&r_c", lang_KRC);
		check_lang ("c++", lang_CPP);
		check_lang ("cplusplus", lang_CPP);
		check_lang ("c-plus-plus", lang_CPP);
#undef check_lang
		else {
		    fprintf (stderr, "%s: unknown language name `%s'\n",
			     whoami, arg);
		    fprintf (stderr, "\tpick one of: C K&R-C\n");
		    exit (1);
		}
	    }
	    else {
		fprintf (stderr, "%s: unknown control argument -`%s'\n",
			 whoami, arg);
		usage ();
	    }
	}
    }
    if (!filename)
	usage ();
    if (!got_language)
	language = lang_KRC;
    else if (language == lang_CPP) {
	fprintf (stderr, "%s: Sorry, C++ support is not yet finished.\n",
		 whoami);
	exit (1);
    }

    p = xmalloc (strlen (filename) + 5);
    strcpy (p, filename);
    filename = p;
    p = strrchr(filename, '/');
    if (p == NULL)
	p = filename;
    else
	p++;
    ename = p;
    len = strlen (ename);
    p += len - 3;
    if (strcmp (p, ".et"))
	p += 3;
    *p++ = '.';
    /* now p points to where "et" suffix should start */
    /* generate new filenames */
    strcpy (p, "c");
    strcpy (c_file, ename);
    *p = 'h';
    strcpy (h_file, ename);
    strcpy (p, "et");

    yyin = fopen(filename, "r");
    if (!yyin) {
	perror(filename);
	exit(1);
    }

    hfile = fopen(h_file, "w");
    if (hfile == NULL) {
	perror(h_file);
	exit(1);
    }
    fprintf (hfile, warning, h_file);

    cfile = fopen(c_file, "w");
    if (cfile == NULL) {
	perror(c_file);
	exit(1);
    }
    fprintf (cfile, warning, c_file);

    /* prologue */
    if (language == lang_C)
	cpp = c_src_prolog;
    else if (language == lang_KRC)
	cpp = krc_src_prolog;
    else
	abort ();
    while (*cpp)
	fputs (*cpp++, cfile);

    /* parse it */
    yyparse();
    fclose(yyin);		/* bye bye input file */

    fputs ("    0\n};\n\n", cfile);
    for (cpp = struct_def; *cpp; cpp++)
	fputs (*cpp, cfile);
    fprintf(cfile,
	    "static const struct error_table et = { text, %d, %d };\n\n",
	    table_number, current);
    fputs("static struct et_list link = { 0, 0 };\n\n",
	  cfile);
    fprintf(cfile, "void initialize_%s_error_table (%s) {\n",
	    table_name, (language == lang_C) ? "void" : "NOARGS");
    fputs("    if (!link.table) {\n", cfile);
    fputs("        link.next = _et_list;\n", cfile);
    fputs("        link.table = &et;\n", cfile);
    fputs("        _et_list = &link;\n", cfile);
    fputs("    }\n", cfile);
    fputs("}\n", cfile);
    fclose(cfile);

    fprintf (hfile, "extern void initialize_%s_error_table ();\n",
	     table_name);
    fprintf (hfile, "#define ERROR_TABLE_BASE_%s (%dL)\n",
	     table_name, table_number);
    /* compatibility... */
    fprintf (hfile, "\n/* for compatibility with older versions... */\n");
    fprintf (hfile, "#define init_%s_err_tbl initialize_%s_error_table\n",
	     table_name, table_name);
    fprintf (hfile, "#define %s_err_base ERROR_TABLE_BASE_%s\n", table_name,
	     table_name);
    fclose(hfile);		/* bye bye include file */

    return 0;
}

int yyerror(s) char *s; {
    fputs(s, stderr);
    fprintf(stderr, "\nLine number %d; last token was '%s'\n",
	    lineno, current_token);

    return 0;
}
